cmd/ferretdb/main.go

```diff
@@ -297,6 +298,11 @@ func run() {
 		return
 	}
 
+	if cli.AppendCoverageMeta {
+		fmt.Println("Append coverage metadata")
+		return
```

Comment:
Um, but then FerretDB does not start?..

Reply:
The way we are getting coverage of integration test cases is
1. We do `task env-up`
2. we do `task env-ferretdb-build` -  which start ferredb server in coverage mode.
3. we run integration tests `task test-integration-pg`

Since we don't have a way to exit the running ferretdb server, we can't get the coverage report (weird, but that is how it is behaving).  There is some things mentioned to cover this area in the actual proposal https://go.googlesource.com/proposal/+/master/design/51430-revamp-code-coverage.md. But it is not clear how to do that.  But what I have observed is if we somehow trigger this method `func WriteMetaDir(dir  string) error` in the `runtime/coverage` package it will append the coverage report into provided directory. So, by doing this `./bind/ferretdb --append-coverage-meta` we are triggering that function through some event ( we are doing some operation and it is exiting after that operation unlike a running server). That operating appends all the covered lines into coverage report

Reply:
> 2. we do `task env-ferretdb-build` -  which start ferredb server in coverage mode.

We want to have that functionality in debug builds of FerretDB (`ferretdb_debug` build tag), but starting FerretDB itself is not needed for integration tests.

> 3. we run integration tests `task test-integration-pg`

And each test itself starts a test-specific in-process FerretDB. For example:
* https://github.com/FerretDB/FerretDB/blob/main/integration/query_test.go#L36
* https://github.com/FerretDB/FerretDB/blob/main/integration/setup/setup.go#L139
* https://github.com/FerretDB/FerretDB/blob/main/integration/setup/listener.go#L182

What that code does is just starts a FerretDB process and then exit before it starts a network listener below. So we might get some coverage with that, but that's a not a coverage of the test-specific FerretDB instances.

Reply:
Also: https://github.com/FerretDB/FerretDB/issues/2102#issuecomment-1616415729

Reply:
can you please elaborate this statement
```
What that code does is just starts a FerretDB process and then exit before it starts a network listener below. So we might get some coverage with that, but that's a not a coverage of the test-specific FerretDB instances.
```
What I understood from the code is `main_test.go` used to call `run` method which starts a Ferretdb server and keeps it running till all the test cases are done and then we have for each test case we start and stop another Ferretdb server. Now we are able to get coverage profile because this multiple Ferretdb server are part of same test binary.

Is my understand correct? If so, why do we need to call `run` method in `main_test` file, is this to make sure no new binary get's create for each test case?

Reply:
Integration tests run `integration/*_test.go` files: https://github.com/FerretDB/FerretDB/blob/48dad105f741a9fefd8b9fcc716c59ff5320a639/Taskfile.yml#L192
They don't run `cmd/ferretdb/main_test.go`. You can confirm that by editing main_test.go to just panic instead and running integration tests.